### PR TITLE
Handle K8s middleware NS record

### PR DESCRIPTION
**Pull request by chrisohaver**
Adds NS record for the dns service name and ip.
For an NS query: The service domain name (per k8s dns spec), and the service's Cluster IP are returned.
Also, an A record request for "ns.dns.zone" will return the service's Cluster IP. (to make the SOA response valid)

### CODE REVIEW

File: middleware/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -92,13 +101,27 @@ func (k *Kubernetes) Services(state request.Request, exact bool, opt middleware.
 		return nil, nil, e
 	}
 
+	if state.Name() == DefaultNSName+"."+r.zone && state.Type() == "A" {
```
Comment (johnbelamaric) [inline middleware/kubernetes/kubernetes.go]: Do we really need to do the additions with "." and r.zone or can we just do strings.Index(state.Name(), DefaultNSName) == 0 ?

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Also you need to go fmt.

File: middleware/kubernetes/kubernetes.go (hunk 1 above)
Comment (miekg) [reply middleware/kubernetes/kubernetes.go]: `DefaultNSName` could also include a `.`

Also: shall we move some of this logic on another file? This one getting big, `ns.go` or something?

File: middleware/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -80,10 +82,17 @@ type recordRequest struct {
 	port, protocol, endpoint, service, namespace, typeName, zone string
 }
 
+type nsRecord struct {
```
Comment (miekg) [inline middleware/kubernetes/kubernetes.go]: Can't we just use `dns.A`?

File: middleware/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -115,6 +138,73 @@ func (k *Kubernetes) recordsForTXT(r recordRequest) ([]msg.Service, error) {
 	return nil, nil
 }
 
+func (k *Kubernetes) CoreDNSRecord() nsRecord {
```
Comment (miekg) [inline middleware/kubernetes/kubernetes.go]: Can there only be 1?

**Review by miekg — CHANGES_REQUESTED** [review]

File: middleware/kubernetes/kubernetes.go (hunk 3 above)
Comment (chrisohaver) [reply middleware/kubernetes/kubernetes.go]: Realistically/Practically, yes.  Which is why I treat it this way.

In theory, there could be multiple services pointing to the coredns pod, and those services may or may not have the DNS port open.  So a complete answer would be to look at each service that points to the pod, and examine the port forwarding in the service and pod configs to see which ones forward dns to coredns.

But in this PR, I am assuming there is one service for the DNS pod, and that service forwards DNS port to the pod.

File: middleware/kubernetes/kubernetes.go (hunk 1 above)
Comment (chrisohaver) [reply middleware/kubernetes/kubernetes.go]: @johnbelamaric : `strings.Index(state.Name(), DefaultNSName) == 0` would falsely catch an A record request for `ns.dns.svc.cluster.local`... which could be a valid domain name for service "ns" in  namespace "dns".

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Handle K8s mi..." ]
>chrisohaver commented on this pull request.
>
>
>
>> @@ -92,13 +101,27 @@ func (k *Kubernetes) Services(state request.Request, exact bool, opt middleware.
> 		return nil, nil, e
> 	}
>
>+	if state.Name() == DefaultNSName+"."+r.zone && state.Type() == "A" {
>
>@johnbelamaric : `strings.Index(state.Name(), DefaultNSName) == 0` would falsely catch an A record request for `ns.dns.svc.cluster.local`... which could be a valid domain name for service "ns" in  namespace "dns".

ah, we also claiming a namespace here. Should we rename 'dns' to
'kube-system'?

Is `r.zone` static, 'cause the above causes garbage for every query hitting this
middleware.

**Comment by chrisohaver** [comment]
> ah, we also claiming a namespace here. Should we rename 'dns' to 'kube-system'?

Ideally, SOA would return the actual domain name of the dns service.  But currently, backend_lookup SOA returns a hard-coded record to `ns.dns.[zone]`, which means that all middleware needs to add handle an A record request for `ns.dns.[zone]`.

> Is `r.zone` static, 'cause the above causes garbage for every query hitting this middleware.

r.zone can be different for each request. A middleware can handle more then one zone, each request is only for one zone.

**Comment by johnbelamaric** [comment]
Still, we can do the Index check and then double-check inside the if with the complete string. That way we avoid the garbage *unless* we enter the if - it's somewhat unlikely that we will see a service named `ns` in a namespace `dns`.

Or maybe we should change the `dns.dns.` hardcoding.

**Comment by chrisohaver** [comment]
> Or maybe we should change the dns.dns. hardcoding.

Yes, IMO we should change SOA to ask the middleware what the record should be... ideally in a way that doest require us to update all other middleware.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Handle K8s mi..." ]
>> ah, we also claiming a namespace here. Should we rename 'dns' to 'kube-system'?
>
>Ideally, SOA would return the actual domain name of the dns service.  But currently, backend_lookup SOA returns a hard-coded record to `ns.dns.[zone]`, which means that all middleware needs to add handle an A record request for `ns.dns.[zone]`.
>
>> Is `r.zone` static, 'cause the above causes garbage for every query hitting this middleware.
>
>r.zone can be different for each request. A middleware can handle more then one zone, each request is only for one zone.

Ack. but we should try to do this with allocating a new string

**Comment by chrisohaver** [comment]
Latest change ...
* Splits NS record functions into a separate file: ns.go
* Reduces heap allocations per kubernetes middleware query. Based on compile time escape analysis - not run time profiling.

File: Makefile.release [hunk 4]
```diff
@@ -38,7 +38,7 @@ NAME:=coredns
 VERSION:=$(shell grep 'coreVersion' coremain/version.go | awk '{ print $$3 }' | tr -d '"')
 ARCH:=$(shell uname -m)
 GITHUB:=coredns
-DOCKER:=coredns
+DOCKER:=chrisohaver
```
Comment (johnbelamaric) [inline Makefile.release]: This file shouldn't be included in the PR.

File: Makefile [hunk 5]
```diff
@@ -15,7 +15,7 @@ deps: core/zmiddleware.go core/dnsserver/zdirectives.go
 	go get -u github.com/golang/lint/golint
 
 .PHONY: check
-check: fmt deps
+check: deps
```
Comment (johnbelamaric) [inline Makefile]: this file shouldn't be in here either

Comment (chrisohaver) [reply Makefile]: ugh... i wonder what else snuck in... My intent was to only commit ns.go and kubernetes.go

**Comment by chrisohaver** [comment]
Looking into the pod lookup related integration test failures.
Pod lookups (both insecure and verified) work when i do manual testing in a real k8s deployment.
Not yet sure why it fails in travis integration.

File: Makefile (hunk 5 above)
Comment (chrisohaver) [reply Makefile]: OK - mystery solved... at some point I accidentally dropped the "docker" target off a  `make -f Makefile.release docker` command ...
`make -f Makefile.release` automatically does a `git commits -am"Release $(VERSION)"` ...

**Comment by johnbelamaric** [comment]
Ah ha!

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Handle K8s mi..." ]
>chrisohaver commented on this pull request.
>
>
>
>> @@ -15,7 +15,7 @@ deps: core/zmiddleware.go core/dnsserver/zdirectives.go
> 	go get -u github.com/golang/lint/golint
>
> .PHONY: check
>-check: fmt deps
>+check: deps
>
>OK - mystery solved... at some point I accidentally dropped the "docker" target off a  `make -f Makefile.release docker` command ...
>`make -f Makefile.release` automatically does a `git commits -am"Release $(VERSION)"` ...

Ai, that's worth fixing

**Comment by johnbelamaric** [comment]
@miekg any thoughts on this comment from earlier:

>> ah, we also claiming a namespace here. Should we rename 'dns' to 'kube-system'?

>Ideally, SOA would return the actual domain name of the dns service. But currently, backend_lookup SOA returns a hard-coded record to ns.dns.[zone], which means that all middleware needs to add handle an A record request for ns.dns.[zone].

I think that's the only outstanding question here.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Handle K8s mi..." ]
>@miekg any thoughts on this comment from earlier:
>
>>> ah, we also claiming a namespace here. Should we rename 'dns' to 'kube-system'?
>
>>Ideally, SOA would return the actual domain name of the dns service. But currently, backend_lookup SOA returns a hard-coded record to ns.dns.[zone], which means that all middleware needs to add handle an A record request for ns.dns.[zone].
>
>I think that's the only outstanding question here.

I think we should not claim another namespace and default to kube-system as that
would interfere least with the rest of the system? WDYT?

**Comment by chrisohaver** [comment]
I think the PR is fine as is in this regard.

`dns` is not a valid object type (valid types are `svc` or `pod`), so there is no chance of namespace collision. `myservice.ns.dns.[zone].` cannot reference a valid kubernetes service.

The only trouble that the hard coded SOA response presents is that every middleware has to implement/handle an A record for that static entry.  This is already added for K8s (in this PR).  The dns service will have two A records...
- one at `ns.dns.zone.`
- the other typically at `kube-dns.kube-system.svc.zone.` 

The PTR record points to `kube-dns.kube-system.svc.zone.`

**Comment by johnbelamaric** [comment]
Ok, so from a k8s perspective there isn't really an issue here. `ns.dns.cluster.local.` doesn't conflict with any other record that k8s middleware should handle. @miekg from a DNS perspective is there an issue with the PTR not pointing to the A record that matches the SOA?

I think we probably just need a separate issue to discuss/decide whether we want CoreDNS to hard code the SOA name and force an A record with that name.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Handle K8s mi..." ]
>Ok, so from a k8s perspective there isn't really an issue here. `ns.dns.cluster.local.` doesn't conflict with any other record that k8s middleware should handle. @miekg from a DNS perspective is there an issue with the PTR not pointing to the A record that matches the SOA?

Ok. Let's use 'dns'. Nothing should check a PTR to a SOA's MNAME (I think)

>I think we probably just need a separate issue to discuss/decide whether we want CoreDNS to hard code the SOA name and force an A record with that name.

Yes.

**Review by johnbelamaric — APPROVED** [review]

**Comment by johnbelamaric** [comment]
@miekg I think this is ready - it doesn't hit the diff coverage number because of improvements to code that is tested by the integration but would require mocking the API to cover it in UT.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Handle K8s mi..." ]
>@miekg I think this is ready - it doesn't hit the diff coverage number because of improvements to code that is tested by the integration but would require mocking the API to cover it in UT.

Ack. I'll give it a review in ~2 hours

File: middleware/kubernetes/controller.go [hunk 6]
```diff
@@ -35,6 +35,14 @@ func (s *storeToNamespaceLister) List() (ns api.NamespaceList, err error) {
 	return ns, nil
 }
 
+type dnsControllerItf interface {
```
Comment (miekg) [inline middleware/kubernetes/controller.go]: Naming: Drop the Itf and just leave it as `dnsController` and rename the type below to `dnsControl`

File: middleware/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -234,7 +248,15 @@ func (k *Kubernetes) parseRequest(lowerCasedName, qtype string) (r recordRequest
 		}
 	}
 	if r.zone == "" {
-		return r, errors.New("zone not found")
+		return r, errZoneNotFound
+	}
+
+	if qtype == "NS" {
```
Comment (miekg) [inline middleware/kubernetes/kubernetes.go]: Most dns type comparisons use dns.TypeXX (`dns.TypeNS`), how did we end up with a string here?

File: middleware/kubernetes/ns.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+		Host: ns.A.String(),
+		Key:  msg.Path(strings.Join([]string{ns.Hdr.Name, r.zone}, "."), "coredns")}
+	*svcs = append(*svcs, s)
+	return nil
+}
+
+// DefaultNSMsg returns an msg.Service representing an A record for
+// ns.dns.[zone] -> dns service ip. This A record is needed to legitimize
+// the SOA response in middleware.NS(), which is hardcoded at ns.dns.[zone].
+func (k *Kubernetes) DefaultNSMsg(r recordRequest) msg.Service {
```
Comment (miekg) [inline middleware/kubernetes/ns.go]: Do these methods all need to be exported?

File: middleware/kubernetes/ns.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	return strings.Index(name, DefaultNSName) == 0 && strings.Index(name, r.zone) == len(DefaultNSName)
+}
+
+type ifAddrsFunc func() ([]net.Addr, error)
+
+func (k *Kubernetes) CoreDNSRecord() dns.A {
+	return k.DoCoreDNSRecord(net.InterfaceAddrs)
+}
+
+func (k *Kubernetes) DoCoreDNSRecord(ifAddrs ifAddrsFunc) dns.A {
```
Comment (miekg) [inline middleware/kubernetes/ns.go]: Drop the `Do' and lowercase `Core`.

Why the `CoreDNSRecord()` -> `DoCoreDNSRecord` indirection? `DoCoreDNSRecord` is useful by itself?

**Review by miekg — COMMENTED** [review]
Looks good, some nits and comments.

File: middleware/kubernetes/ns.go (hunk 9 above)
Comment (chrisohaver) [reply middleware/kubernetes/ns.go]: That was for quick and dirty stubbing.  I have replaced it with a more idiomatic struct wrapper and interface.

File: middleware/kubernetes/controller.go (hunk 6 above)
Comment (chrisohaver) [reply middleware/kubernetes/controller.go]: I did consider naming the interface dnsControllerer. :wink:

**Review by miekg — APPROVED** [review]

**Comment by miekg** [comment]
Test fail, pasting here because I just took a look:
~~~
--- FAIL: TestKubernetesIntegrationFallthrough (1.04s)
	kubernetes_test.go:436: Expected rcode 0 but got 2 for query cluster.local., 2
	kubernetes_test.go:440: Expected 1 answers but got 0 for query cluster.local., 
~~~

**Pull request merged**